packages/flutter_driver/lib/src/common/wait.dart

```diff
@@ -0,0 +1,286 @@
+    };
+  }
+}
+
+/// A condition that waits until no pending frame is scheduled.
+class NoPendingFrameCondition implements WaitCondition {
+  /// Creates a [NoPendingFrameCondition] instance.
+  const NoPendingFrameCondition();
+
+  /// Factory constructor to parse a [NoPendingFrameCondition] instance from the
+  /// given JSON map.
+  factory NoPendingFrameCondition.deserialize(Map<String, dynamic> json) {
+    if (json['conditionName'] != 'NoPendingFrameCondition')
+      throw SerializationException('Error occurred during deserializing the NoPendingFrameCondition JSON string: $json');
+    return const NoPendingFrameCondition();
+  }
+
+  @override
+  bool get condition => !SchedulerBinding.instance.hasScheduledFrame;
+
+  @override
+  Future<void> wait() async {
+    while (!condition) {
+        await SchedulerBinding.instance.endOfFrame;
+    }
+    assert(condition);
+  }
+
+  @override
+  Map<String, String> serialize() {
+    return <String, String>{
+      'conditionName': 'NoPendingFrameCondition',
+    };
+  }
+}
+
+/// A condition that waits until the Flutter engine has rasterized the first frame.
+class FirstFrameRasterizedCondition implements WaitCondition {
+  /// Creates a [FirstFrameRasterizedCondition] instance.
+  const FirstFrameRasterizedCondition();
+
+  /// Factory constructor to parse a [NoPendingFrameCondition] instance from the
+  /// given JSON map.
+  factory FirstFrameRasterizedCondition.deserialize(Map<String, dynamic> json) {
+    if (json['conditionName'] != 'FirstFrameRasterizedCondition')
+      throw SerializationException('Error occurred during deserializing the FirstFrameRasterizedCondition JSON string: $json');
+    return const FirstFrameRasterizedCondition();
+  }
+
+  @override
+  bool get condition => WidgetsBinding.instance.firstFrameRasterized;
+
+  @override
+  Future<void> wait() async {
+    await WidgetsBinding.instance.waitUntilFirstFrameRasterized;
+    assert(condition);
+  }
+
+  @override
+  Map<String, String> serialize() {
+    return <String, String>{
+      'conditionName': 'FirstFrameRasterizedCondition',
+    };
+  }
+}
+
+/// A combined condition that waits until all the given [conditions] are met.
+class CombinedCondition implements WaitCondition {
+  /// Creates a [CombinedCondition] instance with the given list of
+  /// [conditions].
+  ///
+  /// The [conditions] argument cannot be null.
+  const CombinedCondition(this.conditions)
+    : assert(conditions != null);
+
+  /// Factory constructor to parse a [CombinedCondition] instance from the given
+  /// JSON map.
+  factory CombinedCondition.deserialize(Map<String, dynamic> jsonMap) {
+    if (jsonMap['conditionName'] != 'CombinedCondition')
+      throw SerializationException('Error occurred during deserializing the CombinedCondition JSON string: $jsonMap');
+    CombinedCondition combinedCondition;
+    if (jsonMap['conditions'] == null) {
+      combinedCondition = const CombinedCondition(<WaitCondition>[]);
```

Comment:
You could short-circut here and just return const CombinedCondition(<WaitCondition>[]);?

That would also remove the else below, decrease indentation and make things easier to read.